crypto/ctype.c

```diff
@@ -0,0 +1,556 @@
+   /* 92*/ CTYPE_MASK_graph | CTYPE_MASK_print | CTYPE_MASK_punct,
+   /* 93*/ CTYPE_MASK_graph | CTYPE_MASK_print | CTYPE_MASK_punct,
+   /* 94*/ CTYPE_MASK_graph | CTYPE_MASK_print | CTYPE_MASK_punct,
+   /* 95*/ CTYPE_MASK_graph | CTYPE_MASK_print | CTYPE_MASK_punct,
+   /* 96*/ CTYPE_MASK_graph | CTYPE_MASK_print | CTYPE_MASK_punct,
+   /* 97*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print
+           | CTYPE_MASK_xdigit,
+   /* 98*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print
+           | CTYPE_MASK_xdigit,
+   /* 99*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print
+           | CTYPE_MASK_xdigit,
+   /*100*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print
+           | CTYPE_MASK_xdigit,
+   /*101*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print
+           | CTYPE_MASK_xdigit,
+   /*102*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print
+           | CTYPE_MASK_xdigit,
+   /*103*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*104*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*105*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*106*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*107*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*108*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*109*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*110*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*111*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*112*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*113*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*114*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*115*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*116*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*117*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*118*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*119*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*120*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*121*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*122*/ CTYPE_MASK_graph| CTYPE_MASK_lower | CTYPE_MASK_print,
+   /*123*/ CTYPE_MASK_graph | CTYPE_MASK_print | CTYPE_MASK_punct,
+   /*124*/ CTYPE_MASK_graph | CTYPE_MASK_print | CTYPE_MASK_punct,
+   /*125*/ CTYPE_MASK_graph | CTYPE_MASK_print | CTYPE_MASK_punct,
+   /*126*/ CTYPE_MASK_graph | CTYPE_MASK_print | CTYPE_MASK_punct,
+   /*127*/ CTYPE_MASK_cntrl
+};
+#endif
+
+int ossl_ctype_check(int c, int mask)
+{
+    const int max = OSSL_NELEM(ctype_char_map);
+
+#ifdef CHARSET_EBCDIC
+    if (c < 0)
+        c += 256
+#endif
+    return c >= 0 && c < max && (ctype_char_map[c] & mask) != 0;
+}
+
+#ifdef CHARSET_EBCDIC
+static const int case_change = -64;
+#else
+static const int case_change = 32;
+#endif
+
+int ossl_toupper(int c)
+{
+#ifdef CHARSET_EBCDIC
+    if (c < 0)
```

Comment:
> Any (primitive, integral) type of lower conversion rank than 'int' will be promoted to 'int'

I do understand *that*. Point rather is that declaring argument as unsigned char will effectively make caller abstain from sign extension, which would supposedly improve *overall* experience. As already mentioned elsewhere, "char" doesn't actually mean universally signed, and problem in the context is that people tend to have habits. In other words they might expect something that it is not case. And unsigned char declaration should kind of eliminate the ambiguity...